### PR TITLE
added compilation gas cost

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: arwen/context/arwen.go [hunk 1]
```diff
@@ -190,14 +193,21 @@ func (host *vmContext) RunSmartContractCreate(input *vmcommon.ContractCreateInpu
 	return vmOutput, err
 }
 
-func (host *vmContext) computeInitialCreateCost(input *vmcommon.ContractCreateInput) uint64 {
-	storePerByteCost := host.GasSchedule().BaseOperationCost.StorePerByte
-	createCost := host.GasSchedule().ElrondAPICost.CreateContract
+func (host *vmContext) computeInitialCodeCost(
```
Comment (camilbancioiu) [inline arwen/context/arwen.go]: Rename to `deductInitialCodeCost`.

File: arwen/context/arwen.go [hunk 2]
```diff
@@ -138,13 +138,16 @@ func (host *vmContext) RunSmartContractCreate(input *vmcommon.ContractCreateInpu
 	host.scAddress = address
 	host.addTxValueToSmartContract(input.CallValue, address)
 
-	initialCreateCost := host.computeInitialCreateCost(input)
-
-	if input.GasProvided < initialCreateCost {
+	host.vmInput.GasProvided, err = host.computeInitialCodeCost(
+		input.GasProvided,
+		input.ContractCode,
+		host.GasSchedule().ElrondAPICost.CreateContract,
```
Comment (camilbancioiu) [inline arwen/context/arwen.go]: In `CreateNewContract()` below, there is a comment saying "creation costs different for ethereum and elrond". But where is this distinction made?

File: arwen/context/arwen.go [hunk 3]
```diff
@@ -804,6 +840,8 @@ func (host *vmContext) execute(input *vmcommon.ContractCallInput) error {
 	convertedResult := arwen.ConvertReturnValue(result)
 	host.Finish(convertedResult.Bytes())
 
+	totalGasConsumed = input.GasProvided - gasLeft - newInstance.GetPointsUsed()
```
Comment (camilbancioiu) [inline arwen/context/arwen.go]: Variable not used?

File: arwen/context/arwen.go [hunk 4]
```diff
@@ -804,6 +840,8 @@ func (host *vmContext) execute(input *vmcommon.ContractCallInput) error {
 	convertedResult := arwen.ConvertReturnValue(result)
```
Comment (camilbancioiu) [inline arwen/context/arwen.go]: Potential gas leak: my malicious SmartContract could run very expensive code, then trigger a `signalError()`, which will set `host.returnCode` to `vmcommon.UserError`. The lines above (see `if host.returnCode != vmcommon.Ok {...}`) do not deduct gas if this happens.

File: arwen/context/errors.go [hunk 5]
```diff
@@ -11,3 +11,5 @@ var ErrFunctionRunError = errors.New("function run error")
 var ErrFuncNotFound = errors.New("function not found")
 
 var ErrReturnCodeNotOk = errors.New("return not is not ok")
+
+var ErrNotEnoughGas = errors.New("not enough gas")
```
Comment (camilbancioiu) [inline arwen/context/errors.go]: Is this needed? We have `vmcommon.OutOfGas`.

Comment (sasurobert) [reply arwen/context/errors.go]: no. as that is an error code only. here it is the full error.

File: arwen/context/arwen.go (hunk 3 above)
Comment (sasurobert) [reply arwen/context/arwen.go]: used in defer.

File: arwen/context/arwen.go (hunk 2 above)
Comment (sasurobert) [reply arwen/context/arwen.go]: ethereumEI and ElrondEI.

File: arwen/context/arwen.go (hunk 1 above)
Comment (sasurobert) [reply arwen/context/arwen.go]: done.

**Pull request merged**